x/gamm/keeper/pool_service.go

```diff
@@ -200,7 +200,7 @@ func (k Keeper) JoinPoolNoSwap(
 	// check that needed lp liquidity does not exceed the given `tokenInMaxs` parameter. Return error if so.
 	// if tokenInMaxs == 0, don't do this check.
 	if tokenInMaxs.Len() != 0 {
-		if !(neededLpLiquidity.DenomsSubsetOf(tokenInMaxs) && tokenInMaxs.IsAllGTE(neededLpLiquidity)) {
+		if !(neededLpLiquidity.DenomsSubsetOf(tokenInMaxs) && tokenInMaxs.IsAllGTE(neededLpLiquidity) && tokenInMaxs.DenomsSubsetOf(neededLpLiquidity)) {
```

Comment:
It feels that the error message does not apply here:
```
TokenInMaxs is less than the needed LP liquidity to this JoinPoolNoSwap
```

What do you think about having a separate check that produces a relevant error message?

Reply:
Agreed – refactored the new logic to have a new error message that says the following and returns an `ErrDenomNotFoundInPool` error:

```
"TokenInMaxs includes tokens that are not part of the target pool,"
```